### PR TITLE
AVRO-3374: special cases for qualified name

### PR DESCRIPTION
**Pull request by clesaec**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO-3374) issues and references them in the PR title.

### Tests

- [x] My PR adds the following unit tests TestSchema.testParseRecordWithNameAsType and TestSchema.testQualifiedName

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does (no adding functionality nor modified public classes)
  
R:@RyanSkraba

### CODE REVIEW

**Comment by clesaec** [comment]
@RyanSkraba : friendly reminder, if you have time to see this PR.

File: lang/java/avro/src/main/java/org/apache/avro/Schema.java [hunk 1]
```diff
@@ -751,8 +751,32 @@ public void writeName(Names names, JsonGenerator gen) throws IOException {
...
+    private boolean shouldWriteFull(String defaultSpace) {
+      if (space != null && space.equals(defaultSpace)) {
+        try {
+          // name is a 'Type', so namespace must be written (int should return true, Int
+          // should return false).
+          return Type.valueOf(this.name.toUpperCase(Locale.ENGLISH)).name.equals(this.name);
+        } catch (IllegalArgumentException ex) {
+          // namespace can be omitted, (default & name is not a type)
+          return false;
+        }
```
Comment (opwvhk) [inline lang/java/avro/src/main/java/org/apache/avro/Schema.java:774]: This code catches the exception as flow control, which is quite expensive. I think we should prefer a comparison loop instead:
```
      for (Type type : Type.values()) {
        if (type.name.equals(name)) {
          return true;
        }
      }
      return false;
```

Comment (clesaec) [reply lang/java/avro/src/main/java/org/apache/avro/Schema.java:774]: Well, Type.valueOf use an internal hashmap (see Class.enumConstantDirectory method), that will have a direct access to Type values ...
And, as reminded [here](https://medium.com/javarevisited/micro-optimizations-in-java-good-nice-and-slow-enum-261e6f77bd2e), Type.values() method will build a clone of all values.
So, the code i proposed will be faster when exist (true return without exception), but i don't know for exception creation vs Type.values() with clone.
Tell me what you think, i can change code.

Comment (clesaec) [reply lang/java/avro/src/main/java/org/apache/avro/Schema.java]: But anyway, for 14 values, i think you're right, i updated the code.

**Review by RyanSkraba — APPROVED** [review]
LGTM, thanks @clesaec and @opwvhk for working this out!

**Pull request merged**